packages/core/src/modules/job-listing/job-listing-location-repository.ts

```diff
@@ -1,49 +1,64 @@
-  remove(payload: LocationDelete): Promise<LocationSelect>
+  add(payload: JobListingLocationWrite): Promise<JobListingLocation>
+  remove(id: JobListingLocationId): Promise<void>
+  removeByZeroUsage(): Promise<void>
 
-  getAll(): Promise<LocationSelect[]>
-  getByName(name: string): Promise<LocationSelect>
-  getById(id: LocationId): Promise<LocationSelect>
+  getAll(): Promise<JobListingLocation[]>
+  findByName(name: string): Promise<JobListingLocation | null>
+  findById(id: JobListingLocationId): Promise<JobListingLocation | null>
 }
 
 export class JobListingLocationRepositoryImpl implements JobListingLocationRepository {
   constructor(private readonly db: Kysely<Database>) {}
 
-  async getAll(): Promise<LocationSelect[]> {
+  async getAll(): Promise<JobListingLocation[]> {
     const locations = await this.db.selectFrom("jobListingLocation").selectAll().orderBy("name", "asc").execute()
-    return locations
+    return locations.map(mapToJobListingLocation)
+  }
+
+  async add(payload: JobListingLocationWrite): Promise<JobListingLocation> {
+    const location = await this.db
+      .insertInto("jobListingLocation")
+      .values(payload)
+      .returningAll()
+      .executeTakeFirstOrThrow()
+    return mapToJobListingLocation(location)
   }
 
-  async add(payload: LocationInsert): Promise<LocationSelect> {
-    return this.db.insertInto("jobListingLocation").values(payload).returningAll().executeTakeFirstOrThrow()
+  async remove(id: JobListingLocationId): Promise<void> {
+    await this.db.deleteFrom("jobListingLocation").where("id", "=", id).execute()
   }
 
-  async remove(payload: LocationDelete): Promise<LocationSelect> {
-    const location = await this.getById(payload.id)
-    await this.db.deleteFrom("jobListingLocation").where("id", "=", payload.id).execute()
-    return location
+  async findByName(name: string): Promise<JobListingLocation | null> {
+    const location = await this.db
+      .selectFrom("jobListingLocation")
+      .where("name", "=", name)
+      .selectAll()
+      .executeTakeFirst()
+    return location ? mapToJobListingLocation(location) : null
   }
 
-  async getByName(name: string): Promise<LocationSelect> {
-    return this.db.selectFrom("jobListingLocation").where("name", "=", name).selectAll().executeTakeFirstOrThrow()
+  async findById(id: JobListingLocationId): Promise<JobListingLocation | null> {
+    const location = await this.db.selectFrom("jobListingLocation").where("id", "=", id).selectAll().executeTakeFirst()
+    return location ? mapToJobListingLocation(location) : null
   }
 
-  async getById(id: LocationId): Promise<LocationSelect> {
-    return this.db.selectFrom("jobListingLocation").where("id", "=", id).selectAll().executeTakeFirstOrThrow()
+  async removeByZeroUsage(): Promise<void> {
+    await this.db
+      .deleteFrom("jobListingLocation")
+      .where("id", "not in", (eb) => eb.selectFrom("jobListingLocationLink").select("locationId"))
+      .execute()
```

Comment:
Deletes all unused tags

This query should be quick, because it's maintained at every location change